lib/maplibre-compose/src/commonMain/kotlin/dev/sargunv/maplibrecompose/core/expression/Expression.kt

```diff
@@ -1,81 +1,102 @@
-
-    internal fun ofString(string: String): Expression<String> = Expression(string)
-
-    private fun Float.isSmallInt(): Boolean =
-      this >= 0 && this < constSmallInts.size && this.toInt().toFloat() == this
-
-    internal fun ofFloat(float: Float): Expression<Number> {
-      return when {
-        float.isSmallInt() -> constSmallInts[float.toInt()]
-        (float * 20f).isSmallInt() -> constSmallFloats[(float * 20f).toInt()]
-        else -> Expression(float)
-      }
-    }
-
-    @Suppress("UNCHECKED_CAST")
-    internal fun ofDp(dp: Dp): Expression<Dp> = ofFloat(dp.value) as Expression<Dp>
-
-    internal fun ofColor(color: Color): Expression<Color> =
-      when (color) {
-        Color.Transparent -> constTransparent
-        Color.Black -> constBlack
-        Color.White -> constWhite
-        else -> Expression(color)
-      }
-
-    internal fun ofBoolean(bool: Boolean): Expression<Boolean> = if (bool) constTrue else constFalse
-
-    internal fun ofNull(): Expression<Nothing?> = constNull
-
-    internal fun ofOffset(offset: Offset): Expression<Offset> =
-      if (offset == Offset.Zero) constZeroOffset else Expression(offset)
-
-    @Suppress("UNCHECKED_CAST")
-    internal fun ofDpOffset(dpOffset: DpOffset): Expression<DpOffset> =
-      (if (dpOffset == DpOffset.Zero) constZeroOffset
-      else Expression(Offset(dpOffset.x.value, dpOffset.y.value)))
-        as Expression<DpOffset>
-
-    @Suppress("UNCHECKED_CAST")
-    internal fun <T : LayerPropertyEnum> ofLayerPropertyEnum(enum: T): Expression<T> =
-      enum.expr as Expression<T>
-
-    internal fun ofPadding(padding: PaddingValues.Absolute): Expression<PaddingValues.Absolute> =
-      if (padding == ZeroPadding) constZeroPadding else Expression(padding)
-
-    // return Expression<*> because without ["literal" ... ] MapLibre may not treat it as a list
-    internal fun ofList(list: List<Expression<*>>): Expression<*> =
-      Expression<Any?>(list.map { it.value })
-
-    // return Expression<*> because without ["literal" ... ] MapLibre may not treat it as a list
-    internal fun ofMap(map: Map<String, Expression<*>>): Expression<*> =
-      Expression<Any?>(map.entries.associate { (key, value) -> key to value.value })
-  }
+/** Wraps a JSON-like value that represents an expression, typically used for styling map layers. */
+public sealed interface Expression {
+  /** The JSON-like value that backs this expression. */
+  public val value: Any?
 }
+
+/** Represents an expression that resolves to a true or false value. */
+public sealed interface BooleanExpression : Expression
+
+/**
+ * Represents an expression that resolves to any numeric quantity. Corresponds to numbers in the
+ * JSON style spec.
+ */
+public sealed interface ScalarExpression :
+  Expression, MatchableExpression, InterpolateableExpression, ComparableExpression
+
+/** Represents an expression that resolves to an integer quantity. */
+public sealed interface IntScalarExpression : ScalarExpression
```

Comment:
What's the difference between IntScalarExpression and IntExpression?

(Also, shouldn't it be IntegerExpression because that's how the type is named in Kotlin?)

Reply:
Scalar is the super type of dimensionless quantities (number) and those with dimensions (dp). So a DpExpression is a FloatScalarExpression but not a FloatExpression

Reply:
maybe I don't understand what is meant with "dimension". You mean, a unit?

(I remember the words of my physics teacher in school: "What is the result of this equation?" - student: "It's 20" - teacher: "20 what? eggs?")

Reply:
yup exactly. scalars are anything that's represented by a single numeric value.

A scalar with a unit shouldn't be comparable with scalars of other units. Arithmetic like addition and division should match units on both sides, while arithmetic like multiplication and division may have a unit on at most one side. (I still need to fix the functions to support this).

would be nice to somehow write tests for this; in the TypeScript world I often write "tests" that certain code should fail to compile by marking it with `// @ts-expect-error`. Rust has a `compile_fail` attribute in doc tests. Kotlin, to my knowledge, has nothing for this.

Reply:
```
@Test foo() {
  assertFails { throw RuntimeException() }
}
```

Reply:
oh nevermind, you meant fails to compile

Reply:
yup, it's really handy any time you're writing code with complex type checking. comes up often in TS and Rust, not so often in Kotlin but would be handy here

maybe https://github.com/tschuchortdev/kotlin-compile-testing can help